core-bundle/src/Routing/ResponseContext/CoreResponseContextFactory.php

```diff
@@ -85,6 +90,26 @@ public function createContaoWebpageResponseContext(PageModel $pageModel): Respon
             $htmlHeadBag->setMetaRobots($pageModel->robots);
         }
 
+        if ($pageModel->enableCanonical && $pageModel->canonicalLink) {
+            /** @var Controller $controller */
+            $controller = $this->contaoFramework->getAdapter(Controller::class);
+            $url = $controller->replaceInsertTags($pageModel->canonicalLink, false);
```

Comment:
`Controller::replaceInsertTags` is deprecated. Can we use `(new InsertTags())->replace()` here instead?

Reply:
No we can't because we cannot mock that. We will be able to replace this with @ausi's new service he's building in https://github.com/contao/contao/pull/3638